source/common/router/router_ratelimit.cc

```diff
@@ -397,33 +395,55 @@ void RateLimitPolicyEntryImpl::populateLocalDescriptors(
     const std::string& local_service_cluster, const Http::RequestHeaderMap& headers,
     const StreamInfo::StreamInfo& info) const {
   RateLimit::LocalDescriptor descriptor({});
-  bool result =
+  const bool result =
       populateDescriptor(actions_, descriptor.entries_, local_service_cluster, headers, info);
   if (result) {
     descriptors.emplace_back(descriptor);
   }
 }
 
-RateLimitPolicyImpl::RateLimitPolicyImpl()
-    : rate_limit_entries_reference_(RateLimitPolicyImpl::MAX_STAGE_NUMBER + 1) {}
-
-RateLimitPolicyImpl::RateLimitPolicyImpl(
+namespace {
+// Initializes the ``RateLimitPolicyImpl::rate_limit_entries_`` vector.
+std::vector<RateLimitPolicyEntryImpl> initRateLimitEntries(
     const Protobuf::RepeatedPtrField<envoy::config::route::v3::RateLimit>& rate_limits,
-    Server::Configuration::CommonFactoryContext& context, absl::Status& creation_status)
-    : RateLimitPolicyImpl() {
-  // Pre-reserve space in the vector to prevent reallocations.
-  // This ensures that references stored in rate_limit_entries_reference_[stage] remain valid
-  // throughout the lifetime of the ``RateLimitPolicyImpl`` object.
-  rate_limit_entries_.reserve(rate_limits.size());
+    Server::Configuration::CommonFactoryContext& context, absl::Status& creation_status) {
+  std::vector<RateLimitPolicyEntryImpl> entries;
+  entries.reserve(rate_limits.size());
   for (const auto& rate_limit : rate_limits) {
-    rate_limit_entries_.emplace_back(rate_limit, context, creation_status);
-    RETURN_ONLY_IF_NOT_OK_REF(creation_status);
-    const RateLimitPolicyEntryImpl& rate_limit_policy_entry = rate_limit_entries_.back();
-    const uint64_t stage = rate_limit_policy_entry.stage();
-    ASSERT(stage < rate_limit_entries_reference_.size());
-    rate_limit_entries_reference_[stage].emplace_back(std::ref(rate_limit_entries_.back()));
+    entries.emplace_back(rate_limit, context, creation_status);
+    // If the entry is invalid, the function returns no policy entries, and
+    // changes the creation_status to the error.
+    if (!creation_status.ok()) {
+      return {};
+    }
+  }
+  return entries;
+}
+
+// The maximum stage number supported. This value should match the maximum stage number in
+// Json::Schema::HTTP_RATE_LIMITS_CONFIGURATION_SCHEMA and
+// Json::Schema::RATE_LIMIT_HTTP_FILTER_SCHEMA from common/json/config_schemas.cc.
+static constexpr uint64_t MAX_STAGE_NUMBER = 10UL;
+
+// Initializes the ``RateLimitPolicyImpl::rate_limit_entries_reference_`` vector.
+std::vector<std::vector<std::reference_wrapper<const RateLimitPolicyEntry>>>
+initRateLimitEntriesReference(const std::vector<RateLimitPolicyEntryImpl>& rate_limit_entries) {
+  std::vector<std::vector<std::reference_wrapper<const RateLimitPolicyEntry>>> references(
+      MAX_STAGE_NUMBER + 1);
+  for (const auto& entry : rate_limit_entries) {
+    const uint64_t stage = entry.stage();
+    ASSERT(stage < references.size());
```

Comment:
`RELEASE_ASSERT`? Otherwise a configuration that goes this way is a buffer overflow in production. (Is this already validated in config validation so it doesn't get here?)

Reply:
hmm, interesting point...
I [kept it an ASSERT](https://github.com/envoyproxy/envoy/blob/f2a1a61901d7b7ed3a7ced176fd0540859ef5d9c/source/common/router/router_ratelimit.cc#L419), but I agree that it should be something else. I think the config should be rejected in this case, so let me add that with a test.

Reply:
/wait

Reply:
Turns out that this is being [validated by PGV](https://github.com/envoyproxy/envoy/blob/8a60ac8c62d0dae8c2ba0df1402348377cc29421/api/envoy/config/route/v3/route_components.proto#L2275), so I think this should stay an assert (no need to have RELEASE_ASSERT as it doesn't need to be compiled in prod binaries).